packages/x-tree-view/src/internals/plugins/useTreeViewSelection/useTreeViewSelection.ts

```diff
@@ -0,0 +1,210 @@
+      }
+    } else {
+      base.push(next);
+      currentRangeSelection.current.push(current, next);
+    }
+
+    if (params.onNodeSelect) {
+      (params.onNodeSelect as UseTreeViewSelectionDefaultizedParameters<true>['onNodeSelect'])!(
+        event,
+        base,
+      );
+    }
+
+    models.selected.setValue(base);
+  };
+
+  const handleRangeSelect = (
+    event: React.SyntheticEvent,
+    nodes: { start: string; end: string },
+  ) => {
+    let base = (models.selected.value as string[]).slice();
+    const { start, end } = nodes;
+    // If last selection was a range selection ignore nodes that were selected.
+    if (lastSelectionWasRange.current) {
+      base = base.filter((id) => currentRangeSelection.current.indexOf(id) === -1);
+    }
+
+    let range = getNodesInRange(start, end);
+    range = range.filter((node) => !instance.isNodeDisabled(node));
+    currentRangeSelection.current = range;
+    let newSelected = base.concat(range);
+    newSelected = newSelected.filter((id, i) => newSelected.indexOf(id) === i);
+
+    if (params.onNodeSelect) {
+      (params.onNodeSelect as UseTreeViewSelectionDefaultizedParameters<true>['onNodeSelect'])!(
+        event,
+        newSelected,
+      );
+    }
+
+    models.selected.setValue(newSelected);
+  };
+
+  const selectRange = (event: React.SyntheticEvent, nodes: TreeViewItemRange, stacked = false) => {
+    if (params.disableSelection) {
+      return;
+    }
+
+    const { start = lastSelectedNode.current, end, current } = nodes;
+    if (stacked) {
+      handleRangeArrowSelect(event, { start, next: end, current });
+    } else if (start != null && end != null) {
+      handleRangeSelect(event, { start, end });
+    }
+    lastSelectionWasRange.current = true;
+  };
+
+  const rangeSelectToFirst = (event: React.KeyboardEvent<HTMLUListElement>, nodeId: string) => {
+    if (!lastSelectedNode.current) {
+      lastSelectedNode.current = nodeId;
+    }
+
+    const start = lastSelectionWasRange.current ? lastSelectedNode.current : nodeId;
+
+    instance.selectRange(event, {
+      start,
+      end: getFirstNode(instance),
+    });
+  };
+
+  const rangeSelectToLast = (event: React.KeyboardEvent<HTMLUListElement>, nodeId: string) => {
+    if (!lastSelectedNode.current) {
+      lastSelectedNode.current = nodeId;
+    }
+
+    const start = lastSelectionWasRange.current ? lastSelectedNode.current : nodeId;
+
+    instance.selectRange(event, {
+      start,
+      end: getLastNode(instance),
+    });
+  };
+
+  populateInstance<UseTreeViewSelectionSignature<any>>(instance, {
+    isNodeSelected,
+    selectNode,
+    selectRange,
+    rangeSelectToLast,
+    rangeSelectToFirst,
+  });
+
+  return {
+    getRootProps: () => ({
+      'aria-multiselectable': params.multiSelect,
+    }),
+  };
+};
+
+useTreeViewSelection.models = {
+  selected: { controlledProp: 'selected', defaultProp: 'defaultSelected' },
+};
+
+const defaultDefaultSelected: string[] = [];
```

Comment:
Maybe rename it to `DEFAULT_DEFAULT_SELECTED` or just `DEFAULT_SELECTED` in order to avoid confusion of using double `default`. 🙈